drivers/flash/flash_page_layout.c

```diff
@@ -72,6 +72,41 @@ size_t z_impl_flash_get_page_count(const struct device *dev)
 	return count;
 }
 
+int z_impl_flash_get_erase_region(const struct device *dev, off_t offset,
+				  struct flash_erase_region *region)
+{
+	if ((dev == NULL) || (region == NULL) || (offset < 0)) {
+		return -EINVAL;
+	}
+
+	const struct flash_driver_api *api =
+		(const struct flash_driver_api *)dev->api;
+	const struct flash_pages_layout *layout;
+	size_t layout_size;
+
+	api->page_layout(dev, &layout, &layout_size);
+
+	region->start_offset = 0;
+	while (layout_size--) {
+		region->size = layout->pages_size * layout->pages_count;
+
+		if (offset < region->size) {
+			size_t eb_jmp = offset / layout->pages_size;
+
+			region->start_offset += eb_jmp * layout->pages_size;
+			region->size -= eb_jmp * layout->pages_size;
```

Comment:
Assuming that page size is always power of 2, which is across device tree as far as I can tell, this can be replaced with:
```
size_t eb_jmp = offset & ~(layout->pages_size - 1);
region->start_offset += eb_jmp;
region->size -= eb_jmp;
```

Reply:
I agree that page size for physical devices is always a power of 2. To allow support of a virtual device that uses part of the physical block for e.g. an ecc it would be good to also allow non power of 2 pages.

Reply:
I doubt we will leak ecc part to anything above driver and I doubt somebody would make successful direct/random address device with strange block size. It would be pain to make other stuff work with it. Anyway, not insisting, as I have already half-acked the PR as is.

Reply:
Thanks for your comment.

I was thinking about a virtual flash device on top of a physical flash with a power of 2 erase-block-size (e.g. to support wear-levelling or a ecc on a 4kB flash device taking 512 bytes for the wear levelling or ecc, this would then expose 3584 bytes as erase-block-size).

Anyway on the tests I made it does not seem to make any difference in code size if we use the knowledge of power of 2 or not.